src/basic/memory-util.h

```diff
@@ -78,7 +79,17 @@ static inline void* explicit_bzero_safe(void *p, size_t l) {
 void *explicit_bzero_safe(void *p, size_t l);
 #endif
 
+static inline void erase_and_freep(void *p) {
+        void *ptr = *(void**) p;
+
+        if (ptr) {
+                size_t l = malloc_usable_size(ptr);
+                explicit_bzero_safe(ptr, l);
+                free(ptr);
```

Comment:
my initial patch did something like this, but as our earlier use of malloc_usable_size() showed is something memory checkers really really don't like. My earlier patch then tried to use realloc() (like the other time when we used malloc_usable_size()) to extend the memory block to the usable size before erasing it. However during testing I found out thtat in some cases realloc() to malloc_usable_size() of the same memory block actually does move the memory... (at least in valgrind) and if it moves the memory then the excercise is pointless, since we can't actually erase the correct memory that way...

It's why I then came up with the other macro...

Reply:
Hmm. We're don't need to realloc here, so maybe this will be enough. The version of the patch with `malloc_usable_size` is so much shorter and nicer than the other approaches... If we can't make this to work, then I'll try your other idea.

Reply:
So, it appears at least the glibc devs are clear that msan and the other chekcers should be fixed to allow this use, hence I'd say, let's just merge this.

https://sourceware.org/bugzilla/show_bug.cgi?id=24775#c6